src/libraries/System.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/JSProxyContext.cs

```diff
@@ -0,0 +1,609 @@
+        }
+
+        public void FreeJSVHandle(nint jsvHandle)
+        {
+            lock (this)
+            {
+                JSVHandleFreeList.Add(jsvHandle);
+            }
+        }
+
+        // A JSOwnedObject is a managed object with its lifetime controlled by javascript.
+        // The managed side maintains a strong reference to the object, while the JS side
+        //  maintains a weak reference and notifies the managed side if the JS wrapper object
+        //  has been reclaimed by the JS GC. At that point, the managed side will release its
+        //  strong references, allowing the managed object to be collected.
+        // This ensures that things like delegates and promises will never 'go away' while JS
+        //  is expecting to be able to invoke or await them.
+        public IntPtr GetJSOwnedObjectGCHandle(object obj, GCHandleType handleType = GCHandleType.Normal)
+        {
+            if (obj == null)
+            {
+                return IntPtr.Zero;
+            }
+
+            lock (this)
+            {
+                if (ThreadJsOwnedObjects.TryGetValue(obj, out IntPtr gcHandle))
+                {
+                    return gcHandle;
+                }
+
+                IntPtr result = (IntPtr)GCHandle.Alloc(obj, handleType);
+                ThreadJsOwnedObjects[obj] = result;
+                return result;
+            }
+        }
+
+        // TODO unregister and collect pending PromiseHolder also when no C# is awaiting ?
+        public PromiseHolder GetPromiseHolder(nint gcHandle)
+        {
+            PromiseHolder holder;
+            if (IsGCVHandle(gcHandle))
+            {
+                lock (this)
+                {
+                    holder = new PromiseHolder(this, gcHandle);
+                    ThreadJsOwnedHolders.Add(gcHandle, holder);
+                }
+            }
+            else
+            {
+                holder = (PromiseHolder)((GCHandle)gcHandle).Target!;
+            }
+            return holder;
+        }
+
+        public unsafe void ReleaseJSOwnedObjectByGCHandle(nint gcHandle)
+        {
+            PromiseHolder? holder = null;
+            lock (this)
+            {
+                if (IsGCVHandle(gcHandle))
+                {
+                    if (ThreadJsOwnedHolders.Remove(gcHandle, out holder))
+                    {
+                        holder.GCHandle = IntPtr.Zero;
+                    }
+                    else
+                    {
+                        throw new InvalidOperationException("ReleaseJSOwnedObjectByGCHandle expected in ThreadJsOwnedHolders");
+                    }
+                }
+                else
+                {
+                    GCHandle handle = (GCHandle)gcHandle;
+                    var target = handle.Target!;
+                    if (target is PromiseHolder holder2)
+                    {
+                        holder = holder2;
+                        holder.GCHandle = IntPtr.Zero;
```

Comment:
this looks like a potential GCHandle leak to me

Reply:
I made it readonly